presto-main/src/main/java/com/facebook/presto/metadata/BuiltInFunctionNamespaceManager.java

```diff
@@ -785,8 +801,13 @@ public FunctionMetadata getFunctionMetadata(FunctionHandle functionHandle)
             throw e;
         }
         SqlFunction function = functionKey.getFunction();
+        FunctionImplementationType implementationType = function instanceof SqlInvokedFunction ? SQL : BUILTIN;
```

Comment:
If we add `getLanguage` to the `SqlFunction` interface, we can either make `BUILTIN` an enum of `Language`, or make `getLanguage` returns Optional. I'm not a fan for either of those choices.

"builtin" is not really a "language". It's a way of implementation. IMO, language should be derived from the tuple (SqlFunction.getClass, SqlInvokedFunction.getLanguage).

Also, it's undefined for `SqlFunction` to have a language `SQL`, but having `getLanguage` method would allow so, and that's extra check rather than fewer check.

Having `getLanguage` method only affects this line. The 2 checks are line 810 and line 420-423, but making this change won't let those checks go away.

Reply:
For the builtin java functions, the language could be JAVA, for the builtin sql functions the language would be SQL. Presto can run JAVA and SQL language natively. JAVA => BUILTIN, SQL => SQL. And in the future when we introduce other languages, those would have an implementation type REMOTE.

Reply:
If you wonder what about Hive java functions, the language for that will be HIVE 😂

Reply:
My point is not about what enum names to use.

Adding `getLanguage` to SqlFunction leaves a hole in the API. User will be allowed to define a SQL-invoked function (either through `CREATE FUNCTION` statement, or through plugin) with language `Java`. Similarly, SqlScalarFunction can have `language SQL`. In both cases, functions can be created but fails with internal error during execution.

Reply:
It's not a hole. CREATE FUNCTION would not support language Java. Language is a rightful metadata that every function should have. Please separate implementation from concept. Conceptually it's valid that every function should have a language. The the implementation type is a derived information from language, because the engine supports some languages not others.

Reply:
Practically you can override this in `BuiltInFunction` as final so all BuiltInFunctions has Language Java.

Reply:
> Language is a rightful metadata that every function should have.
We need to clarity the meaning of the word here.

Currently, the enum class `Language` represents a property of a SQL-invoked function, as defined in the Spec. "Language" in this context is either "SQL", if the SQL-invoked Function is a SQL-invoked Regular Function, or the programming language in which the external routine is implemented.

What you've proposed is to "repurpose" or "extends" the enum "Language" to represent a different concept. I think repurposing something is total fine but the exact new definition of the enum `Language` is unknown to me. Maybe you can clarity this?

"If" this new definition `Language` still represent the programming language in which a Presto function is implemented, then a user-defined SQL function implemented in "Java" should also have a language "Java", not "Hive". Hive is not language anyway, `HQL` is, but is not applicable in this context here.

Consider those 3 categories of function:
- A scalar function implemented in Presto or Presto Plugin, using the annotation framework that Presto can understand. (SqlScalarFunction)
- A scalar function implemented in Hive code base, using the framework that Hive can understand.
- A scalar function implemented as a SQL-invoked external routine, with the routine name point to a file containing Java code. (SqlInvokedFunction).
All 3 of them are implemented in the Java programming language.

I would say, even if we generalize the definition of language from `SqlInvokedFunction` to all `SqlFunction` in general, all 3 cases should have language `Java`. The suggestion that you made about assigning `Language.Java` to case 1, and `Language.Hive` to case 2 seems to me that you're looking to convey extra information with the "language" concept or the "Language" enum. The extra information is "what framework that the Java implementation adheres to" - does it use the Presto annotation (Can be parsed by Presto to create a `SqlScalarFunction`), or is it a Java class enclosing a Java method like the Hive-style.

Going back to what you say "Language is a rightful metadata that every function should have." I agree if by saying "language", you actually mean the programming language in which the function is implemented, and in all of those 3 cases, "Java". I don't agree if the "language" you're talking about also conveys the extra information about the framework/assumption that the Java implementation uses, because
1.  That should not be called "language".
2. That can be derived from the class object.
3. That's a special case for language Java because Presto and Hive happens to be implemented in Java, and therefore it should not be packed together with "SQL", "Python", "JS", etc., the one that actually only means programming language.

> Please separate implementation from concept
I gave `CREATE FUNCTION dev.leiqing.rand() LANGUAGE JAVA` example, not to discuss how we're going to implement `CREATE FUNCTION`, but to align us on what the expected correct behavior of this query should be. Now that you confirmed that this should not work.

Reply:
@rongrong Sorry about the long post here. Just to make things clear, the first part is talking about the `Language` enum, and the second part is talking about `getLanguage` function.

Reply:
Yes, I'm suggesting extending the meaning of `Language`. You've already touched it. Builtin Java functions, Hive java functions, or java functions implemented with another API, are all JAVA functions. So we either repurpose the `Language` to give them different values, or we need to introduce another type of metadata (and it's not implementation type). I'd choose just repurpose the language, which is more compatible with SQL standard than introducing a new metadata.

Reply:
Discussed offline. We'll have a single enum that can be used to derive all execution information, language, and framework with a language. `FunctionMetadata` will contains this enum, and the enum will be serialized between Presto server and remote UDF server (when we have that in the future). The enum does not need to be exposed in `presto-spi`, and enum `Language` will be kept as its original definition in the spec.

`BuiltInFunctionNamespaceManager.getFunctionMetadata` will be the single place to figure out the value of this enum from other information.

I think we can do this in a separate PR given that this enum is not exactly too useful until we have remote execution cases, and this PR has already been throughly reviewed.